File: CSW/JavaScript/220920_최소원_감시피하기.js
var fs = require('fs');
var input = fs.readFileSync(`/dev/stdin`).toString().trim().split('\n');

let N = parseInt(input[0]);
const board = [];
const teacher = [];
const empty = [];
const dirY = [-1,0,1,0];
const dirX = [0,1,0,-1];
let possible = 0;

for(let i=0; i<N; i++){
    const col = input[1+i].trim().split(' ');
    for(let j=0; j<col.length; j++){
        if(col[j]==='X'){
            empty.push([i,j]);
        }else if(col[j]==='T'){
            teacher.push([i,j]);
        }
    }
    board.push(col);
}

function find(){

    for(let i=0; i<teacher.length; i++){
        let ty = teacher[i][0];
        let tx = teacher[i][1];
        for(let j=0; j<4; j++){
            let dy = ty;
            let dx = tx;
            while(1){
                dy += dirY[j];
                dx += dirX[j];
                if(dy<0||dx<0|| dy>=N||dx>=N || board[dy][dx] === 'O') break;
                // 학생들이 감시 피하는거 실패
                if(board[dy][dx] === 'S'){
                    return false;
                }
            }
        }
    }
    // 감시피하기 성공
    return true;
}

function dfs(level){
    if(level === 3){
        if(find()) possible = 1;
        return;
    }

    for(let i=0; i<empty.length; i++){
        let y = empty[i][0];
        let x = empty[i][1];
        if(board[y][x] ==='X'){
            board[y][x] = 'O';
            dfs(level+1);
            board[y][x] = 'X';
        }
    }

}


dfs(0);

if(possible) console.log("YES");
else console.log("NO");

File: CSW/C,C++/220920_최소원_감시피하기.cpp
#include <iostream>
#include <vector>
using namespace std;

int N, possible;
vector<pair<int, int>> teacher;
vector<pair<int, int>> emp;
char map[6][6];
int dirY[4] = { -1,0,1,0 };
int dirX[4] = { 0,1,0,-1 };

bool find() {

	for (int i = 0; i < teacher.size(); i++) {
		int ty = teacher[i].first;
		int tx = teacher[i].second;
		for (int j = 0; j < 4; j++) {
			int dy = ty;
			int dx = tx;
			while(1) {
				dy += dirY[j];
				dx += dirX[j];
				if (dy < 0 || dx < 0 || dy >= N || dx >= N || map[dy][dx] == 'O') break;
				if (map[dy][dx] == 'S') return false;
			}
		}
	}
	return true;
	
}

void dfs(int level) {
	if (level == 3) {
		if (find()) possible = 1;
		return;
	}

	for (int i = 0; i < emp.size(); i++) {
		int ey = emp[i].first;
		int ex = emp[i].second;
		if (map[ey][ex] == 'X') {
			map[ey][ex] = 'O';
			dfs(level + 1);
			map[ey][ex] = 'X';
		}
	}
}

int main() {
	cin >> N;

	for (int i = 0; i < N; i++) {
		for (int j = 0; j < N; j++) {
			cin >> map[i][j];
			if (map[i][j] == 'X') emp.push_back({ i,j });
			else if (map[i][j] == 'T') teacher.push_back({ i,j });
		}
	}


	dfs(0);

	if (possible) cout << "YES";
	else cout << "NO";

	return 0;
}